ignite/metrics/ssim.py

```diff
@@ -158,17 +159,31 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
             )
 
         channel = y_pred.size(1)
-        if len(self._kernel.shape) < 4:
-            self._kernel = self._kernel.expand(channel, 1, -1, -1).to(device=y_pred.device)
+
+        self._kernel = self._kernel.expand(channel, 1, -1, -1)
+        kernel = self._kernel
+
+        if y_pred.device != self._kernel.device:
+            warnings.warn(
+                "The metric's device should be the same than your update tensors. See SSIM() device argument.",
+                RuntimeWarning,
+            )
```

Comment:
I have an impression that we can skip the warning.
I imagine the following situation, user create SSIM with device=cpu and does NN computations on cuda, so that y and y_pred would be on cuda. When eval part would be executed he will see this warning but it is legit to have the setup they have.

Reply:
What would be the benefits of having the metric on CPU now that if y_pred is on GPU and kernel on CPU, kernel will be pushed to the GPU to perform the computation?

I could rephrase the warning to underline the lower performance when the two tensors are on different devices.

Reply:
device arg for the metric in general is the device on which we store intermediate results and it is not computation device. In SSIM it happens that kernel is stored on metric device as well. But for conv2d computations it would be better to have it on the faster device (or cpu for cpu-only hardware).
Right now I think that actually, previous "weird" code was doing right thing about converting `self._kernel` to `y_pred` device once vs right now we do that on each iteration. What do you think ?
In general we also would like that user do not bother about the best config and by default it should be the best.

Reply:
That was my commit comment. But now, because we are trying to do the "right thing" by putting all the tensors on GPU, there is the edge case where (y_pred1, y1) are on GPU but (y_pred2, y2) are on CPU.
If we move self._kernel (which is typically way lighter than y_pred) on GPU, then the second iteration will move y_pred on GPU (which will be longer, but maybe not as long as computing the SSIM metric entirely on CPU).

Reply:
I guess now we just have to define what's the best config !

Reply:
We can add also `.to(device, non_blocking=True)`, to hide a bit the device sync.

> here is the edge case where (y_pred1, y1) are on GPU but (y_pred2, y2) are on CPU.

I think this is very edge case and IMO it is ok to put y_pred and y from cpu to cuda in this case.


> I guess now we just have to define what's the best config !

I would say let's compute all on cuda if available even if metric device is cpu

Reply:
> device("cuda") Vs device("cuda:0").

this can be done as 
```
x = torch.tensor(0.0, device="cuda:0")
assert x.device.type == "cuda"
```